### PR TITLE
WIP: Default page templates

### PR DESCRIPTION
**Pull request by lennertdaniels**
I took a stab at implementing default page templates as indicated in #1803.

### What I have implemented:
- There's a book-level setting to pick a default template
- A new page automatically copies the template's html on creation
- A new guests page also has the template applied to it
- Deleted pages don't work as a template anymore
- When deleting a page that is used as a default template somewhere, there's a warning on the delete confirmation screen

image

image

### What I haven't fully considered or implemented yet:
- Any permission/scope problems
- If there's a difference in handling HTML vs. markdown templates
- A better selector to pick the default template (pagination, AJAX search) like the one in the book.edit sidebar (template-manager)
- Should other entities also support a default template? _(User, Shelf, Chapter)_ If so, which one has priority when multiple default templates are chosen on different levels?
- Applying a template from a URL parameter (@cod3monk and #3794)

### Random small fixes I came across:
- Autofocus on the name field on guest-create.blade
- Template Manager: fixed cancel icon placement

If anyone with more thorough knowledge of the codebase would like to shine some light onto this PR, that would be great.

### CODE REVIEW

**Comment by lennertdaniels** [comment]
@ssddanbrown Would you mind taking a look if you find the time? Thanks.

**Comment by ssddanbrown** [comment]
Hi @lennertdaniels,
Sorry for my lack of response so far, I plan to review pending PRs after finishing up the next release.
I remember doing a quick scan across the PR code when opened, and from what I can remember it was all looking really good, with maybe just needing some thought into ensuring the selection is scalable/efficient. Of course I need to look into this to properly review your code and considerations.
The one bit of feedback I will provide now is to not worry about any optional sub-features at this time (query param, other entities). I prefer to keep implementation/scope narrow and only expand on proven requirement/demand, to keep the platform surface area reasonable. 

Should review towards the end of the month unless things go further wrong (Current release cycle is being a challenge).

**Comment by lennertdaniels** [comment]
@ssddanbrown absolutely fine, take your time. I just wasn't sure wether I had to add some tag/reviewer/assignee or something  to this PR. :)

I'll take your quick feedback into consideration and work on it a little more over the coming weeks.

File: app/Entities/Repos/PageRepo.php
Comment (ssddanbrown) [inline app/Entities/Repos/PageRepo.php:156]: This currently ignores permissions. I think we'll need to respect page permissions, so view access is required for the template to be used, otherwise other side-effects will be introduced (Image access permissions for example).

Comment (ssddanbrown) [inline app/Entities/Repos/PageRepo.php:156]: We do need to support markdown also here, and probably the original used editor.
I suspect it should just be a case of also copying the `markdown` and `editor` values from the original page model.

File: app/Http/Controllers/BookController.php
Comment (ssddanbrown) [inline app/Http/Controllers/BookController.php:86]: Right now, this'll be loading in all template pages, with all their properties (including html, markdown & text) into memory for use in a list. Probably fine for the majority of use-cases but I'm uncomfortable with the scalability of this.

It's a pain but ideally template selection would be done via front-end querying, much like what's shown when using the "Move" or "Copy" interfaces, to prevent requiring load of all options on form load.

File: app/Http/Controllers/BookController.php
Comment (ssddanbrown) [inline app/Http/Controllers/BookController.php:108]: Right now there's nothing stopping a user from setting a page template they don't have access to. 
We probably should have some light validation to check the user can view and that the template page is marked as a template. 

This exists validation check could also be abused to check the existence of a certain page id what the user may not have access to. Not sure how well we currently defend against that in other areas of the app, but probably would be something I'd attempt to avoid where possible.

File: app/Http/Controllers/BookController.php
Comment (ssddanbrown) [inline app/Http/Controllers/BookController.php:165]: As per create method comment.

File: app/Http/Controllers/BookController.php
Comment (ssddanbrown) [inline app/Http/Controllers/BookController.php:187]: As per create method comment.

Comment (ssddanbrown) [inline app/Http/Controllers/BookController.php:187]: Currently, if the template option is removed when editing a book the column value will end up as `0`. Since nulls are supported in this column, i think the value should end up `null` instead.

File: app/Http/Controllers/PageController.php
Comment (ssddanbrown) [inline app/Http/Controllers/PageController.php:269]: Variable name does not follow project conventions, but use may be redundant (lang file comments).

File: resources/lang/en/entities.php
Comment (ssddanbrown) [inline resources/lang/en/entities.php:195]: I don't think we need to show the count here (Which again can complicate matters when permissions are in play) but I just think we need the user to be aware of the consequences, something like:

> This page is in active use as a book default page template. Such books will no longer have a page default template assigned after this page is deleted.

File: resources/lang/en/entities.php
Comment (ssddanbrown) [inline resources/lang/en/entities.php:333]: This should probably include a note on permissions, advising that the page template will only be used if the page creator has view permission for the page template.

File: resources/views/entities/template-manager.blade.php
Comment (ssddanbrown) [inline resources/views/entities/template-manager.blade.php:2]: There's no newlines in the source string so this can be a simple non-escaped template tag. If there are multiple clear lines I try to split the translation itself these days.

File: resources/views/pages/parts/template-manager.blade.php
Comment (ssddanbrown) [inline resources/views/pages/parts/template-manager.blade.php:17]: I'd generally advise slipping small fixes into larger works otherwise they get blocked by the larger feature and/or conflict if the fix gets applied in the meantime.

**Review by ssddanbrown — CHANGES_REQUESTED** [review]
Thanks for this @lennertdaniels, All was clear and seemed to function well. 
Just raised my thoughts/findings from a deeper dive into the code.
Feel free to query or ask for clarification/assistance/guidance for any of my notes.

Ideally we'd have testing to cover the functionality additions here also, at least to cover the main functionality. Happy to help or advise on that inf you're unfamiliar with testing.

**Comment by ssddanbrown** [comment]
Hi @lennertdaniels,
Would you be looking to address the review points above yourself? Or do you need a certain level of support/advise? Or would you prefer for me to carry this on from this point?

**Comment by lennertdaniels** [comment]
Hi @ssddanbrown,

I'm slowly chipping away at your remarks whenever I find some spare time :) Will update the PR in due time.
The feedback was very useful, thanks!

**Comment by melat0nin** [comment]
Hi @lennertdaniels, have you managed to make any progress on this PR? It's a feature I'd really love to have in BookStack :)

**Comment by ssddanbrown** [comment]
Hi @lennertdaniels,
Would you be okay for me to take this on and finish things off from this point?

**Comment by lennertdaniels** [comment]
@ssddanbrown please do! 
I wasn't able to dedicate a lot of time to this PR, sorry. I'm excited to see this feature come alive :)

**Comment by ssddanbrown** [comment]
Alright, I'll assign this to take on for the next release.
Please keep your fork & branch, and this PR up and active in the meantime until this is merged.

> I wasn't able to dedicate a lot of time to this PR, sorry.

No reason to be sorry, thanks for contributing and getting this started!

**Comment by ssddanbrown** [comment]
Okay, now all merged with follow-up work continued in #4721.
This will be part of the next feature release.

Thanks again for getting this started @lennertdaniels!

**Pull request merged**
